functorch/experimental/_cond.py

```diff
@@ -193,18 +200,41 @@ def _has_potential_branch_input_alias(branch, inputs):
     except Exception as e:
         raise e
 
-    input_storages = set()
-    for node in gm.graph.nodes:
-        if node.op == "placeholder":
-            input_storages.add(StorageWeakRef(node.meta['val']._typed_storage()))
-        if node.op == "output":
-            for out in node.args:
-                out_storage = StorageWeakRef(out.meta["val"]._typed_storage())
-                if out_storage in input_storages:
-                    return True
+    def _detect_input_alias(gm):
+        input_storages = set()
+        for node in gm.graph.nodes:
+            if node.op == "placeholder":
+                input_storages.add(StorageWeakRef(node.meta['val']._typed_storage()))
+            if node.op == "output":
+                for out in node.args:
+                    out_storage = StorageWeakRef(out.meta["val"]._typed_storage())
+                    if out_storage in input_storages:
+                        return True
+
+        for _, module in gm.named_children():
+            if isinstance(module, torch.fx.GraphModule) and _detect_input_alias(module):
+                return True
+
+        return False
+
+    return _detect_input_alias(gm)
+
+
+@cond.py_impl(DispatchKey.Functionalize)
+def cond_func(pred, true_fn, false_fn, inputs):
+    unwrapped_inputs = _unwrap_all_tensors_from_functional(inputs, reapply_views=torch._C._functionalization_reapply_views_tls())
+    unwrapped_pred = _unwrap_all_tensors_from_functional(pred, reapply_views=torch._C._functionalization_reapply_views_tls())
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.Functionalize))
```

Comment:
These type of guards need to be wrapped in try-finally, otherwise if an exception happens they won't be deleted (the exception traceback may hold onto the object).

i.e.
```py
guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.Functionalize))
try:
    ...
finally:
    del guard
```

Reply:
Yep did.